client/pov-recovery/src/lib.rs

```diff
@@ -435,28 +442,40 @@ where
 }
 
 /// Returns a stream over pending candidates for the parachain corresponding to `para_id`.
-fn pending_candidates(
-	relay_chain_client: impl RelayChainInterface,
+async fn pending_candidates(
+	relay_chain_client: impl RelayChainInterface + Clone,
 	para_id: ParaId,
-) -> impl Stream<Item = (CommittedCandidateReceipt, SessionIndex)> {
-	relay_chain_client.import_notification_stream().filter_map(move |n| {
-		let res = relay_chain_client
-			.candidate_pending_availability(&BlockId::hash(n.hash), para_id)
-			.and_then(|pa| {
-				relay_chain_client
-					.session_index_for_child(&BlockId::hash(n.hash))
-					.map(|v| pa.map(|pa| (pa, v)))
-			})
-			.map_err(|e| {
-				tracing::error!(
-					target: LOG_TARGET,
-					error = ?e,
-					"Failed fetch pending candidates.",
-				)
-			})
-			.ok()
-			.flatten();
-
-		async move { res }
-	})
+) -> RelayChainResult<impl Stream<Item = (CommittedCandidateReceipt, SessionIndex)>> {
+	let import_notification_stream = relay_chain_client.import_notification_stream().await?;
+
+	let filtered_stream = import_notification_stream.filter_map(move |n| {
+		let client_for_closure = relay_chain_client.clone();
+		async move {
+			let block_id = BlockId::hash(n.hash());
+			let pending_availability_result = client_for_closure
+				.candidate_pending_availability(&block_id, para_id)
+				.await
+				.map_err(|e| {
+					tracing::error!(
+						target: LOG_TARGET,
+						error = ?e,
+						"Failed to fetch pending candidates.",
+					)
+				});
+			let session_index_result =
+				client_for_closure.session_index_for_child(&block_id).await.map_err(|e| {
+					tracing::error!(
+						target: LOG_TARGET,
+						error = ?e,
+						"Failed to fetch session index.",
+					)
+				});
+
+			session_index_result
+				.and_then(|v| Ok(pending_availability_result?.map(|pa| (pa, v))))
```

Comment:
```suggestion
				.and_then(|v| pending_availability_result.map(|pa| (pa, v)))
```

Reply:
This suggestion does not work here as it changes the return type. However, I tried to improve the readability of this section